mitodl/ocw-studio
# Course publish UI

#### Pre-Flight checklist


#### What are the relevant tickets?
Fixes #621 

#### What's this PR do?
Implements a UI for publishing courses based on the design in #621:
- The buttons are now placed in a drawer
- There is an indicator for unsaved changes
- There is info about the last time a site was published
- There is an error message if something went wrong

There is also code added to set `has_unpublished_live` and `has_unpublished_draft` to `True` on `Website` when a `WebsiteStarter` or `WebsiteContent` is saved. This marks the website as needing a republish.

Note that we don't do this when `Website` is updated since that's a little trickier to implement and I didn't see a use case in the UI. If we have one we can implement it in a future PR. 

#### How should this be manually tested?
 - First, set up site publishing if you haven't already. Let me know if you need to do this and I can walk you through it
 - Click the publish button. It should open a drawer and there should be two options for staging and production. Both options should show an indication for unpublished content
 - Click the publish button under either draft or production. The page should refresh, closing the drawer.
 - Click the publish button to open the drawer again. You should not see the an indication that there is unpublished content. If you click on the button, nothing should happen. (The button will turn green on hover but it is actually disabled.)
 - Edit or create some content, for example in the site metadata or on a page. Click save, then click the publish button to open the drawer. You should now see an indication that there is unpublished content.
 - In the backend, add "raise Exception()" in the first line of `WebsiteViewSet.publish` in `websites/views.py`. Open the drawer again by clicking the publish button, then click on Production and click Publish. You should see a red error message appear.


#### Screenshots (if appropriate)
![Screenshot from 2021-10-05 14-34-15](https://user-images.githubusercontent.com/863262/136083748-73d50468-19ab-4ea2-a373-ddf703a709da.png)
![Screenshot from 2021-10-05 14-37-08](https://user-images.githubusercontent.com/863262/136083802-93126457-4c0d-45b9-9674-0569bbb8dfaa.png)
![Screenshot from 2021-10-05 14-41-29](https://user-images.githubusercontent.com/863262/136083808-5873b894-b226-4413-8db9-184f54933d89.png)
